sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -25,6 +25,11 @@ public static T EnsureCompleted<T>(this Task<T> task)
         {
 #if DEBUG
             VerifyTaskCompleted(task.IsCompleted);
+#else
+            if (HasSynchronizationContext())
```

Comment:
Not sure this is worth doing. Might break customers apps that worked just fine before.

Reply:
It may for some custom non-blocking sync context. In that case, they will report a bug and use `Task.Run` as a workaround while we fix it. However, if we don't throw in this case, the most common sync context scenario of interaction with UI will hang without any clue for the user why it has happened.

Reply:
That is of course if we have any bugs like that at all (this exception should be thrown only if we have sync code that waits on incomplete task while should only get the result of completed).